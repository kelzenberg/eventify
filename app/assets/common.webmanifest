{
    "icons": [
        { "src": "/assets/Favicon_192.png", "type": "image/png", "sizes": "192x192" },
        { "src": "/assets/Favicon_512.png", "type": "image/png", "sizes": "512x512" }
    ]
}