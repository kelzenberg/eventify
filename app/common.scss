@use "sass:map";
@import "./bootstrap.custom.scss";

body {
    margin: 0px;
    font-family: 'Athiti', sans-serif;
}

.fancy-background {
    position: fixed;
    top: 0px;
    left: 0px;
    right: 0px;
    bottom: 0px;
    z-index: -5;
    overflow: hidden;
    background: linear-gradient(160deg, #F6CAA2, #F2B680);

    &::before {
        content: "";
        position: absolute;
        z-index: 0;
        top: 0px;
        left: 0px;
        width: 200vw;
        height: 30vh;
        transform: rotateZ(-20deg);
        transform-origin: 0% 100%;
        background: #f5ceaa;
        filter: blur(5px);
    }

    &::after {
        content: "";
        position: absolute;
        z-index: 0;
        bottom: 0px;
        right: 0px;
        width: 200vw;
        height: 30vh;
        transform: rotateZ(-20deg);
        transform-origin: 100% 0%;
        background: #F6CAA2;
        filter: blur(5px);
    }
}

.iconBox {
    display: inline-block;
    position: relative;
    border-radius: 13px;
    border: 1px solid map.get($theme-colors, "primary");
    width: 50px;
    height: 50px;
    background-repeat: no-repeat;
    background-position: 50% 50%;
    vertical-align: middle;

    img {
        width: 100%;
        height: 100%;
    }

    @each $color, $value in $theme-colors {
        &.iconBox-#{$color} {
            border-color: $value;
        }
    }

    &.iconBox-removable:hover::after {
        content: "";
        position: absolute;
        top: 0px;
        left: 0px;
        right: 0px;
        bottom: 0px;
        border-radius: 13px;
        background-color: rgba(255, 255, 255, 0.5);
        background-image: url("/assets/icons/remove.svg");
        background-repeat: no-repeat;
        background-position: 50% 50%;
        pointer-events: none;
    }
}

.hover-accent:hover {
    background-color: #ececec;
}