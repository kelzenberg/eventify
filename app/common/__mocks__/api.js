const api = jest.requireActual("../api");

const UnexpectedServerError = api.default.UnexpectedServerError;

const authenticate = jest.fn().mockResolvedValue({});
const getUserInfo = jest.fn().mockResolvedValue({});
const register = jest.fn().mockResolvedValue({});
const completeRegistration = jest.fn().mockResolvedValue({});
const getAllUserEvents = jest.fn().mockResolvedValue({});
const createEvent = jest.fn().mockResolvedValue({});
const getEvent = jest.fn().mockResolvedValue({});
const saveEvent = jest.fn().mockResolvedValue({});
const leaveEvent = jest.fn().mockResolvedValue({});
const inviteToEvent = jest.fn().mockResolvedValue({});
const bounceFromEvent = jest.fn().mockResolvedValue({});
const addExpenseSharingModule = jest.fn().mockResolvedValue({});
const addPaymentToExpenseSharing = jest.fn().mockResolvedValue({});
const deletePaymentFromExpenseSharing = jest.fn().mockResolvedValue({});

export {
    authenticate,
    getUserInfo,
    register,
    completeRegistration,
    getAllUserEvents,
    createEvent,
    getEvent,
    saveEvent,
    leaveEvent,
    inviteToEvent,
    bounceFromEvent,
    addExpenseSharingModule,
    addPaymentToExpenseSharing,
    deletePaymentFromExpenseSharing,
    UnexpectedServerError
};
