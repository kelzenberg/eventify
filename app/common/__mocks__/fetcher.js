const originalFetcher = jest.requireActual("../fetcher");

const request = jest.fn();

export default {
    FetchError: originalFetcher.default.FetchError,
    request: request,
    status: originalFetcher.default.status
}