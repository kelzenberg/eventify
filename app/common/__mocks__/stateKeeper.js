const stateKeeper = jest.requireActual("../stateKeeper");

export const UserContext = stateKeeper.UserContext;
export const isAuthenticated = jest.fn();
export const setCredentials = jest.fn();
export const getAuthenticationToken = jest.fn();
export const getUserInfo = jest.fn();
export const maybeUpdateUserInfo = jest.fn();
export const clearLogin = jest.fn();