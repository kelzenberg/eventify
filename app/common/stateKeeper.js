import React from 'react';

export const UserContext = React.createContext(null);

export function isAuthenticated() {
    return localStorage.getItem("USER_TOKEN") != null;
}

export function getDisplayName() {
    return "Username";
}

export function setCredentials(token, userInfo) {
    localStorage.setItem("USER_TOKEN", token);
}

export function getAuthenticationToken() {
    return localStorage.getItem("USER_TOKEN");
}

// getUserInfo returns the stored user information or dummy data if none is set
export function getUserInfo() {
    try {
        let data = JSON.parse(localStorage.getItem("USER_INFO"));
        if (data == null) throw new Error();
        return data;
    } catch(e) {
        return {
            "id": "",
            "createdAt": "2021-06-02T00:06:45.528+00:00",
            "updatedAt": "2021-06-02T00:06:45.528+00:00",
            "email": "",
            "displayName": "",
            "authRole": ""
        };
    }
}

// maybeUpdateUserInfo takes a user info structure and checks if it matches the stored information.
// If the informations match it does nothing and returns false.
// If the new user information differs from the stored one it saves the new data and returns true.
export function maybeUpdateUserInfo(newUserInfo) {
    if(localStorage.getItem("USER_INFO") === null) {
        setUserInfo(newUserInfo);
        return true;
    }
    let oldUserInfo = getUserInfo();
    for(let key in newUserInfo) {
        if(newUserInfo[key] !== oldUserInfo[key]) {
            setUserInfo(newUserInfo);
            return true;
        }
    }
    return false;
}

function setUserInfo(userInfo) {
    localStorage.setItem("USER_INFO", JSON.stringify(userInfo));
}

export function clearLogin() {
    localStorage.removeItem('USER_TOKEN');
    localStorage.removeItem('USER_INFO');
}