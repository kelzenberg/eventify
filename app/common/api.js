import fetcher from "./fetcher";
import * as stateKeeper from "./stateKeeper";

export class UnexpectedServerError extends Error {
    constructor(msg, code) {
        super(msg);
    }
}

export function authenticate(email, password) {
    return new Promise((resolve, reject) => {
        fetcher.request({
            method: "POST",
            path: "authenticate",
            expect: "json",
            body: {"email": email, "password": password},
            allowedStatuses: [fetcher.status.unauthorized]
        })
        .then(([responseData, rawResponse]) => {
            if(rawResponse.status === fetcher.status.unauthorized) {
                reject();
            } else {
                stateKeeper.setCredentials(responseData.token);
                resolve();
            }
        })
        .catch(err => {
            reject(new UnexpectedServerError(err));
        })
    });
}

export function getUserInfo() {
    return fetcher.request({method: "GET", path: "me", expect: "json"}).then(([response]) => response);
}

export function register(email, password, displayName) {
    let body = {
        "email": email,
        "password": password,
        "displayName": displayName
    };
    return fetcher.request({method: "POST", path: "register", expect: "json", body: body})
    .then(([responseData, rawResponse]) => {
        stateKeeper.setCredentials(responseData.token);
        return Promise.resolve();
    })
}

export function logout() {
    console.warn("TODO: api logout");
}

// =========================================== EVENTS =========================================== //

export function getAllUserEvents() {
    return fetcher.request({method: "GET", path: "me/events", expect: "json"}).then(([response]) => response);
}

export function getEvent(eventID) {
    return fetcher.request({method: "GET", path: `events/${eventID}`, expect: "json"}).then(([response]) => response);

    return Promise.resolve({
        "id": "22f7c93e-cfdc-491f-baa2-732fa298d8f6",
        "createdAt": "2021-06-02T15:42:16.478+00:00",
        "updatedAt": "2021-06-02T15:42:16.478+00:00",
        "title": "TestEvent 1",
        "description": "This is a test description",
        "startedAt": "2021-12-24T13:33:37.111+00:00",
        "endedAt": null,
        "expenseSharingModules": [
            {
                "id": "0f994c11-58ba-4718-ae18-5fe099177329",
                "createdAt": "2021-06-02T15:42:40.985+00:00",
                "updatedAt": "2021-06-02T15:42:40.985+00:00",
                "title": "Expense Sharing Module 1",
                "description": "This is a test description",
                "payments": [
                    {
                        "id": "e2ba702a-3713-4804-9b67-de5f9f99b3a8",
                        "createdAt": "2021-06-02T15:43:05.301+00:00",
                        "updatedAt": "2021-06-02T15:43:05.301+00:00",
                        "title": "Payment 1",
                        "amount": 13.37,
                        "payer": {
                            "id": "1798a949-1f72-45ff-b83c-d8e796add4cc",
                            "createdAt": "2021-05-29T17:01:31.716+00:00",
                            "updatedAt": "2021-05-29T17:01:31.716+00:00",
                            "email": "admin@test.de",
                            "displayName": "Admin",
                            "authRole": "ADMIN"
                        },
                        "shareType": "FIXED"
                    }
                ]
            }
        ]
    });
}

export function saveEvent(event) {
    return fetcher.request({method: "PUT", path: `events/${event.id}`, expect: "json"}).then(([response]) => response);
}