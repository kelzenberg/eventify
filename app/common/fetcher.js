import * as stateKeeper from "./stateKeeper";

const apiHost = "https://localhost:8443/";

class FetchError extends Error {
    constructor(msg, code) {
        super(msg);
        this.code = code;
    }
}

function request(options) {
    if(typeof options.body == "object") {
        options.body = JSON.stringify(options.body);
    }
    var additionalHeaders = {};
    if(options.headers !== undefined) {
        additionalHeaders = options.headers
    }
    if(options.parameters != undefined) {
        options.path += "?" + new URLSearchParams(options.parameters).toString();
    }
    let authorization = null;
    if(stateKeeper.isAuthenticated()) {
        authorization = `Bearer ${stateKeeper.getAuthenticationToken()}`;
    }
    return new Promise(function(resolve, reject){
        fetch(apiHost + options.path, {
            credentials: 'omit',
            body: options.body,
            headers: {
                'accept': 'application/json',
                'content-type': 'application/json',
                'authorization': authorization,
                ...additionalHeaders
            },
            method: options.method,
            mode: 'cors',
            redirect: 'follow',
            referrer: 'no-referrer',
            keepalive: options.keepalive === true
        })
        // Check for errors
        .then(response => {
            if(response.ok || (options.allowedStatuses != undefined && options.allowedStatuses.includes(response.status)))Â {
                return Promise.resolve(response)
            } else if(response.status >= 400) {
                return Promise.reject(new FetchError(`request returned with status code ${response.status}`, response.status));
            }
        })
        // Decode content
        .then(response => {
            var contentType = response.headers.get("content-type");
            if(contentType == null) {
                if(options.expect != undefined && options.expect != "empty") {
                    console.warn(`fetcher: got response with content type '${contentType}', but expected '${options.expect}'.`);
                    return Promise.reject(new FetchError(`Invalid content-type '${contentType}', but expected '${options.expect}'`));
                }
                return Promise.resolve([null, response]);
            }
            if(contentType.includes("application/json")) {
                if(options.expect != undefined && options.expect != "json") {
                    console.warn(`fetcher: got response with content type '${contentType}', but expected '${options.expect}'.`);
                    return Promise.reject(new FetchError(`Invalid content-type '${contentType}', but expected '${options.expect}'`));
                }
                return response.json().then(data => {
                    return [data, response];
                })
            } else if(contentType.includes("text/html")) {
                if(options.expect != undefined && options.expect != "text") {
                    console.warn(`fetcher: got response with content type '${contentType}', but expected '${options.expect}'.`);
                    return Promise.reject(new FetchError(`Invalid content-type '${contentType}', but expected '${options.expect}'`));
                }
                return response.text().then(data => {
                    return [data, response];
                })
            }
        })
        // Parse data
        .then(([data, response]) => {
            resolve([data, response]);
        })
        // Handle remaining errors
        .catch(error => {
            if(error instanceof FetchError)
                reject(error);
            else {
                console.log(error);
                reject(new FetchError("fetcher encountered an unexpected error: " + error.message, -100))
            }
        });
    });
}

const status = {
    continue: 100,
    switchingProtocols: 101,
    processing: 102,
    earlyHints: 103,
    ok: 200,
    created: 201,
    accepted: 202,
    nonAuthorativeInformation: 203,
    noContent: 204,
    resetContent: 205,
    partialContent: 206,
    alreadyReported: 208,
    imUsed: 226,
    multipleChoice: 300,
    movedPermanently: 301,
    found: 302,
    seeOther: 303,
    notModified: 304,
    useProxy: 305,
    temporaryRedirect: 307,
    permanentRedirect: 308,
    badRequest: 400,
    unauthorized: 401,
    paymentRequired: 402,
    forbidden: 403,
    notFound: 404,
    methodNotAllowed: 405,
    notAcceptable: 406,
    proxyAuthenticationRequired: 407,
    requestTimeout: 408,
    conflict: 409,
    gone: 410,
    lengthRequired: 411,
    preconditionFailed: 412,
    payloadTooLarge: 413,
    uriTooLong: 414,
    unsupportedMediaType: 415,
    requestedRangeNotSatisfiable: 416,
    expectationFailed: 417,
    imATeapot: 418,
    misdirectedRequest: 421,
    unprocessedEntity: 422,
    locked: 423,
    failedDependency: 424,
    tooEarly: 425,
    upgradeRequired: 426,
    preconditionRequired: 428,
    tooManyRequests: 429,
    requestHeaderFieldsTooLarge: 431,
    unavailableForLegalReasons: 451,
    internalServerError: 500,
    notImplemented: 501,
    badGateway: 502,
    serviceUnavailable: 503,
    gatewayTimeout: 504,
    httpVersionNotSupported: 505,
    variantAlsoNegotiates: 506,
    insufficientStorage: 507,
    loopDetected: 508,
    notExtended: 510,
    networkAuthenticationRequired: 511
};

export default {
    FetchError,
    request,
    status
}