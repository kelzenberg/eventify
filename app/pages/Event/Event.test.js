import React from 'react';
import {fireEvent, render, act, waitFor } from '@testing-library/react';
import "@testing-library/jest-dom/extend-expect";
import { Router, Route } from "react-router-dom";
import { createMemoryHistory } from "history";
import update from 'immutability-helper';
import EventPage from "./Event";
import * as api from "../../common/api";
import Timespan from "../../components/Timespan/Timespan";
import fetcher from '../../common/fetcher';
import Title from '../../components/Title/Title';
import { InfoDialog } from '../../components/Dialog/Dialog';
import { UserContext } from '../../common/stateKeeper';

const ModuleCard = EventPage.__RewireAPI__.__get__("ModuleCard");
const ModuleList = EventPage.__RewireAPI__.__get__("ModuleList");
const Details = EventPage.__RewireAPI__.__get__("Details");

jest.mock("../../common/api");
jest.mock("../../components/Header/Header");
jest.mock("../../components/Title/Title");
jest.mock("../../components/Timespan/Timespan");
jest.mock("../../components/Dialog/Dialog");

const userInfo = {
    id: "localUserID",
    displayName: "Local User",
    email: "one@two.three",
    eventRole: "ORGANISER"
};

const event = {
    id: "123456789",
    title: "Event Title",
    startedAt: new Date("2021-02-27T23:38:43.000Z"),
    endedAt: new Date("2021-06-27T23:38:43.000Z"),
    amountOfUsers: 5,
    description: "Event Description",
    users: [userInfo],
    expenseSharingModules: [
        {
            id: "987654321",
            title: "Expense Sharing Module 1",
            description: "Expense Sharing Description 1",
            payments: []
        }, {
            id: "5ab282ff-3ce8-44b3-bc94-ade7d8f10b30",
            title: "Expense Sharing Module 2",
            description: "Expense Sharing Description 2",
            payments: []
        }
    ]
};

describe('ModuleCard', () => {
    test('Normal', async () => {
        let Child = jest.fn(() => null);
        let { getByText } = render(
            <ModuleCard htmlID="" icon="icon.svg" title="Module Name">
                <Child/>
                <span>Content Stuff</span>
            </ModuleCard>
        );

        expect(Child).toHaveBeenCalled();
        let title = getByText("Module Name");
        expect(title).toBeVisible();

        fireEvent.click(title); // collapse
        await waitFor(() => expect(getByText("Content Stuff")).not.toBeVisible());
    });
});

describe('ModuleList', () => {
    let ModuleCardMock;
    let ModuleCreatorMock;
    let ExpenseSharingModuleMock;
    let ChecklistModuleMock;

    beforeAll(() => { // run once before any test is executed
        // mock Event and NewEvent because we don't want to test their behaviour in here
        ModuleCardMock = jest.fn((props) => <div>Mocked Event Card "{props.title}": {props.children}</div>);
        ModuleCreatorMock = jest.fn((props) => <div>Mocked Module Creator</div>);
        ExpenseSharingModuleMock = jest.fn((props) => <span>Mocked Expense Sharing Module</span>);
        ExpenseSharingModuleMock.icon = "Mocked Expense Sharing Module Icon";
        ChecklistModuleMock = jest.fn((props) => <span>Mocked Checklist Module</span>)
        ChecklistModuleMock.icon = "Mocked Checklist Module Icon";
        EventPage.__Rewire__("ModuleCard", ModuleCardMock);
        EventPage.__Rewire__("ModuleCreator", ModuleCreatorMock);
        EventPage.__Rewire__("ExpenseSharingModule", ExpenseSharingModuleMock);
        EventPage.__Rewire__("ChecklistModule", ChecklistModuleMock);
    });

    afterAll(() => { // run once after all tests where executed
        EventPage.__ResetDependency__("ModuleCard");
        EventPage.__ResetDependency__("ModuleCreatorMock");
        EventPage.__ResetDependency__("ExpenseSharingModule");
        EventPage.__ResetDependency__("ChecklistModule");
    });

    afterEach(() => { // run after each test
        ModuleCardMock.mockClear();
        ModuleCreatorMock.mockClear();
        ExpenseSharingModuleMock.mockClear();
        ChecklistModuleMock.mockClear();
    });


    test('Empty', () => {
        let changeHandler = jest.fn();
        let eventWithoutModules = update(event, {expenseSharingModules: {$set: []}});
        let { container, getByText } = render(
            <ModuleList onEventChanged={changeHandler} event={eventWithoutModules}/>
        );
        expect(getByText("new module", {exact: false})).toBeVisible();
        expect(getByText(/no.*modules/i, {exact: false})).toBeVisible();
        expect(ModuleCardMock).not.toHaveBeenCalled();
        expect(container).toMatchSnapshot();

        expect(ModuleCreatorMock).toHaveBeenCalledWith(expect.objectContaining({
            show: false,
        }), expect.anything());
    });

    test('Module Creator', () => {
        let changeHandler = jest.fn();
        let eventWithoutModules = update(event, {expenseSharingModules: {$set: []}});
        let { getByText } = render(
            <ModuleList onEventChanged={changeHandler} event={eventWithoutModules}/>
        );
        // click button to create new module
        fireEvent.click(getByText("new module", {exact: false}));
        expect(ModuleCreatorMock).toHaveBeenCalledWith(expect.objectContaining({
            show: true,
            onHide: expect.any(Function),
            eventID: event.id,
            onModuleAdded: expect.any(Function)
        }), expect.anything());

        // try callback to add a new module
        let creatorProps = ModuleCreatorMock.mock.calls[0][0];
        creatorProps.onModuleAdded("expenseSharingModules", "New Fake Module");
        expect(changeHandler).toHaveBeenCalled();
        expect(changeHandler.mock.calls[0][0].expenseSharingModules).toContain("New Fake Module");
        
        // hide the creator again
        act(() => {
            creatorProps.onHide();
        });
        expect(ModuleCreatorMock).toHaveBeenCalledWith(expect.objectContaining({
            show: false,
        }), expect.anything());
    });

    test('Expense Sharing Module', () => {
        let changeHandler = jest.fn();
        let { getByText } = render(
            <ModuleList onEventChanged={changeHandler} event={event}/>
        );
        expect(getByText("new module", {exact: false})).toBeVisible();

        // check module card
        expect(ModuleCardMock).toHaveBeenCalledWith(expect.objectContaining({
            htmlID: expect.any(String),
            icon: ExpenseSharingModuleMock.icon,
            title: event.expenseSharingModules[0].title
        }), expect.anything());
        expect(ModuleCardMock).toHaveBeenCalledWith(expect.objectContaining({
            htmlID: expect.any(String),
            icon: ExpenseSharingModuleMock.icon,
            title: event.expenseSharingModules[1].title
        }), expect.anything());
        
        // check module itself
        expect(ExpenseSharingModuleMock).toHaveBeenCalledWith(expect.objectContaining({
            moduleData: event.expenseSharingModules[0],
            event: event,
            htmlID: expect.any(String)
        }), expect.anything());
        expect(ExpenseSharingModuleMock).toHaveBeenCalledWith(expect.objectContaining({
            moduleData: event.expenseSharingModules[1],
            event: event,
            htmlID: expect.any(String)
        }), expect.anything());
    });
});

describe('Details', () => {
    test('Normal', () => {
        let changeHandler = jest.fn();
        let { rerender, getByText, getByLabelText } = render(
            <Details event={event} onChange={changeHandler} editing={false}/>
        );

        expect(changeHandler).not.toHaveBeenCalled();
        
        expect(getByText(event.description, {ignore: "textarea"})).toBeVisible();
        expect(Timespan).toHaveBeenCalledWith(expect.objectContaining({
            from: event.startedAt,
            to: event.endedAt,
            editing: false
        }), expect.anything());

        Timespan.mockClear();

        // Editing
        rerender( <Details event={event} onChange={changeHandler} editing={true}/> )

        expect(Timespan).toHaveBeenCalledWith(expect.objectContaining({
            from: event.startedAt,
            to: event.endedAt,
            editing: true
        }), expect.anything());

        fireEvent.change(getByLabelText("Name"), {target: {value: "New Event Name"}});
        expect(changeHandler).toHaveBeenCalledWith(expect.objectContaining({
            title: "New Event Name"
        }));

        fireEvent.change(getByLabelText("Description"), {target: {value: "New Event Description"}});
        expect(changeHandler).toHaveBeenCalledWith(expect.objectContaining({
            description: "New Event Description"
        }));

        let timespanArgs = Timespan.mock.calls[0][0];

        let newStartDate = new Date("2020-02-03T01:02:03.000Z");
        timespanArgs.onChangeFrom(newStartDate);
        expect(changeHandler).toHaveBeenCalledWith(expect.objectContaining({
            startedAt: newStartDate
        }));

        let newEndDate = new Date("2023-02-03T01:02:03.000Z");
        timespanArgs.onChangeTo(newEndDate);
        expect(changeHandler).toHaveBeenCalledWith(expect.objectContaining({
            endedAt: newEndDate
        }));
    });
});

describe('EventPage', () => {
    let DetailsMock;
    let SmallMembersMock;
    let ModuleListMock;
    let FullMembersMock;

    beforeAll(() => { // run once before any test is executed
        // mock Event and NewEvent because we don't want to test their behaviour in here
        DetailsMock = jest.fn((props) => <div>Mocked Details</div>);
        SmallMembersMock = jest.fn((props) => <div>Mocked Small Members</div>);
        ModuleListMock = jest.fn((props) => <div>Mocked Module List</div>);
        FullMembersMock = jest.fn((props) => <div>Mocked Full Members</div>)
        EventPage.__Rewire__("Details", DetailsMock);
        EventPage.__Rewire__("SmallMembers", SmallMembersMock);
        EventPage.__Rewire__("ModuleList", ModuleListMock);
        EventPage.__Rewire__("FullMembers", FullMembersMock);
    });

    afterAll(() => { // run once after all tests where executed
        EventPage.__ResetDependency__("ModuleCard");
        EventPage.__ResetDependency__("ModuleCreatorMock");
        EventPage.__ResetDependency__("ExpenseSharingModule");
        EventPage.__ResetDependency__("ChecklistModule");
    });

    afterEach(() => { // run after each test
        api.getEvent.mockClear();
        api.saveEvent.mockClear();
        InfoDialog.mockClear();
        DetailsMock.mockClear();
        SmallMembersMock.mockClear();
        ModuleListMock.mockClear();
        FullMembersMock.mockClear();
    });

    test('Unknown Event', async () => {
        api.getEvent.mockRejectedValueOnce(new fetcher.FetchError("Event not Found", fetcher.status.notFound));
        const history = createMemoryHistory();
        history.push('/you/event/' + event.id);

        let { findByText } = render(
            <Router history={history}>
                <Route path="/you/event/:eventID">
                    <UserContext.Provider value={userInfo}>
                        <EventPage/>
                    </UserContext.Provider>
                </Route>
            </Router>
        );

        await findByText(/event.*not.*exist/i);
        expect(api.getEvent).toHaveBeenCalledWith(event.id);
    });

    test('API Error', async () => {
        api.getEvent.mockRejectedValueOnce(new Error("Mocked API Error"));
        const history = createMemoryHistory();
        history.push('/you/event/' + event.id);

        let { findByText } = render(
            <Router history={history}>
                <Route path="/you/event/:eventID">
                    <UserContext.Provider value={userInfo}>
                        <EventPage/>
                    </UserContext.Provider>
                </Route>
            </Router>
        );

        await findByText(/internal error/i);
        expect(api.getEvent).toHaveBeenCalledWith(event.id);
    });

    test('Loading', async () => {
        api.getEvent.mockReturnValueOnce(new Promise(() => {}));
        const history = createMemoryHistory();
        history.push('/you/event/' + event.id);

        let { container } = render(
            <Router history={history}>
                <Route path="/you/event/:eventID">
                    <UserContext.Provider value={userInfo}>
                        <EventPage/>
                    </UserContext.Provider>
                </Route>
            </Router>
        );

        expect(Title).toHaveBeenCalledWith(expect.objectContaining({title: expect.stringContaining("Loading")}), expect.anything())
        await waitFor(() => expect(api.getEvent).toHaveBeenCalledWith(event.id));
    });

    test('Normal', async () => {
        api.getEvent.mockResolvedValueOnce(event);
        const history = createMemoryHistory();
        history.push('/you/event/' + event.id);

        let { getByText } = render(
            <Router history={history}>
                <Route path="/you/event/:eventID">
                    <UserContext.Provider value={userInfo}>
                        <EventPage/>
                    </UserContext.Provider>
                </Route>
            </Router>
        );

        // wait for the event to have loaded
        await waitFor(() => 
            expect(Title).toHaveBeenCalledWith(expect.objectContaining({title: event.title}), expect.anything())
        );

        // snapshot of title content
        expect(Title.mock.calls[Title.mock.calls.length-1][0].children).toMatchSnapshot();

        // Details
        expect(DetailsMock).toHaveBeenCalledWith(expect.objectContaining({
            event: event,
            editing: false,
            onChange: expect.any(Function)
        }), expect.anything());

        // Event Members in Sidebar
        expect(SmallMembersMock).toHaveBeenCalledWith(expect.objectContaining({
            event: event,
            onEventChanged: expect.any(Function),
            setVisibleContent: expect.any(Function)
        }), expect.anything());

        // Module List
        expect(ModuleListMock).toHaveBeenCalledWith(expect.objectContaining({
            event: event,
            onEventChanged: expect.any(Function),
            setVisibleContent: expect.any(Function),
            onErrorMessage: expect.any(Function)
        }), expect.anything());

        // Full Member List
        expect(FullMembersMock).toHaveBeenCalledWith(expect.objectContaining({
            event: event,
            onEventChanged: expect.any(Function),
            setVisibleContent: expect.any(Function),
            onErrorMessage: expect.any(Function)
        }), expect.anything());

        // Editing
        fireEvent.click(getByText(/edit/i));

        let detailProps = getPropsOfLastCall(DetailsMock);
        expect(detailProps.editing).toBeTruthy();
        DetailsMock.mockClear();

        // change the title
        act(() => {
            detailProps.onChange(update(event, {title: {$set: "New and different title for the event"}}));
        })
        expect(DetailsMock).toHaveBeenCalled();
        detailProps = getPropsOfLastCall(DetailsMock);
        expect(detailProps.event.title).toBe("New and different title for the event");
        DetailsMock.mockClear();

        // abort editing and check if the old title is back
        fireEvent.click(getByText(/cancel/i));
        expect(DetailsMock).toHaveBeenCalled();
        detailProps = getPropsOfLastCall(DetailsMock);
        expect(detailProps.editing).toBeFalsy();
        expect(detailProps.event.title).toBe(event.title);

        // enable editing again
        fireEvent.click(getByText(/edit/i));

        detailProps = getPropsOfLastCall(DetailsMock);
        expect(detailProps.editing).toBeTruthy();

        // change the title again
        act(() => {
            detailProps.onChange(update(event, {title: {$set: "New and different title for the event"}}));
        });
        expect(DetailsMock).toHaveBeenCalled();
        detailProps = getPropsOfLastCall(DetailsMock);
        expect(detailProps.event.title).toBe("New and different title for the event");


        // now save
        let rejectAPI;
        api.saveEvent.mockImplementation(() => new Promise((resolve, reject) => rejectAPI = reject));

        fireEvent.click(getByText(/save/i));
        await waitFor(() => expect(getByText("Saving...")).toBeVisible());
        // now abort the pending request
        rejectAPI(new Error("Mocked API Error"));
        await waitFor(() => expect(getByText("Saving...")).not.toBeVisible());
        // check that the error is shown to the user via the dialog
        expect(InfoDialog).toHaveBeenCalledWith(expect.objectContaining({
            show: true,
            title: expect.stringMatching(/not.*saved/i)
        }), expect.anything());
    

        // close the dialog and make sure it disappears
        let dialogProps = getPropsOfLastCall(InfoDialog);
        InfoDialog.mockClear();
        act(() => dialogProps.onHide());
        expect(InfoDialog).toHaveBeenCalledWith(expect.objectContaining({
            show: false,
        }), expect.anything());

        // save but this time the api works
        // but it returns a different title that the one we sent to make sure
        // the version from the api is used.
        api.saveEvent.mockResolvedValueOnce(update(event, {title: {$set: "New Title from the API"}}));
        fireEvent.click(getByText(/save/i));

        await waitFor(() =>
            expect(DetailsMock).toHaveBeenCalledWith(expect.objectContaining({
                event: expect.objectContaining({title: "New Title from the API"})
            }), expect.anything())
        );

        // now we test direct updating of the event state
        act(() => getPropsOfLastCall(ModuleListMock).onEventChanged(update(event, {title: {$set: "Directly updated title"}})));
        await waitFor(() => expect(Title).toHaveBeenCalledWith(expect.objectContaining({
            title: "Directly updated title"
        }), expect.anything()));
    });

    test('Leave', async () => {
        api.getEvent.mockResolvedValueOnce(event);
        const history = createMemoryHistory();
        history.push('/you/event/' + event.id);

        let { getByText } = render(
            <Router history={history}>
                <Route path="/you/event/:eventID">
                    <UserContext.Provider value={userInfo}>
                        <EventPage/>
                    </UserContext.Provider>
                </Route>
            </Router>
        );

        // wait for component to be finished
        await waitFor(() => 
            expect(Title).toHaveBeenCalledWith(expect.objectContaining({title: event.title}), expect.anything())
        );

        // try to leave but an api error occurs
        api.leaveEvent.mockRejectedValueOnce(new Error("Mocked API Error"));
        let warnSpy = jest.spyOn(console, "warn"); // prevent console.warn from being called
        warnSpy.mockImplementation(() => {});
        fireEvent.click(getByText("...", {selector: "button"})); // expand drop down
        fireEvent.click(getByText("Leave"));
        await waitFor(() => expect(api.leaveEvent).toHaveBeenCalled());
        // check if the dialog is shown
        expect(InfoDialog).toHaveBeenCalledWith(expect.objectContaining({
            title: expect.stringMatching(/not.*leave/i)
        }), expect.anything())
        warnSpy.mockRestore();

        // close the dialog and try again, this time succeeding
        act(() => getPropsOfLastCall(InfoDialog).onHide());
        api.leaveEvent.mockResolvedValueOnce(null);
        fireEvent.click(getByText("...", {selector: "button"})); // expand drop down
        // await waitFor(() => expect(getByText("Leave")).toBeVisible());
        fireEvent.click(getByText("Leave"));
        await waitFor(() => expect(api.leaveEvent).toHaveBeenCalled());
        expect(history.location.pathname).toMatch(/.*you$/);
    });
});

function getPropsOfLastCall(mockFn){
    return mockFn.mock.calls[mockFn.mock.calls.length-1][0];
}