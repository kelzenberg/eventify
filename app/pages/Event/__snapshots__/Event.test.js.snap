// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EventPage Normal 1`] = `
Array [
  <Link
    to="/you/events"
  >
    <button
      className="btn btn-outline-primary"
      hidden={false}
    >
      GO BACK
    </button>
  </Link>,
  <button
    className="btn btn-outline-secondary"
    hidden={false}
    onClick={[Function]}
  >
    EDIT
  </button>,
  <button
    className="btn btn-outline-secondary"
    hidden={true}
    onClick={[Function]}
  >
    CANCEL
  </button>,
  <button
    className="btn btn-primary"
    hidden={true}
    onClick={[Function]}
  >
    SAVE
  </button>,
  <button
    className="btn btn-primary"
    disabled={true}
    hidden={true}
  >
    <span
      className="pe-2"
    >
      Saving...
    </span>
    <span
      aria-hidden="true"
      className="spinner-border spinner-border-sm"
      role="status"
    />
  </button>,
  <DropdownButton
    as={
      Object {
        "$$typeof": Symbol(react.forward_ref),
        "defaultProps": Object {
          "role": "group",
          "vertical": false,
        },
        "render": [Function],
      }
    }
    title="..."
    variant="outline-secondary"
  >
    <DropdownItem
      as={
        Object {
          "$$typeof": Symbol(react.forward_ref),
          "render": [Function],
        }
      }
      disabled={false}
      onClick={[Function]}
    >
      Leave
    </DropdownItem>
  </DropdownButton>,
]
`;

exports[`ModuleList Empty 1`] = `
<div>
  <h4
    class="text-center"
  >
    This event does not have any modules yet.
  </h4>
  <div
    aria-label="Add new Item"
    class="rounded border-dashed border-gray w-100 p-3 text-center text-primary"
    role="button"
  >
    <img
      alt=""
      class="pe-2"
      src="/assets/icons/add.svg"
      style="vertical-align: sub;"
    />
    ADD NEW MODULE
  </div>
  <div>
    Mocked Module Creator
  </div>
</div>
`;
